src/us/kbase/workspace/database/mongo/ObjectLister.java

```diff
@@ -119,39 +134,52 @@ private DBObject buildProjection(final ResolvedListObjectParameters params) {
 	 */
 	private DBObject buildSortSpec(final ResolvedListObjectParameters params) {
 		final DBObject sort = new BasicDBObject();
-		if (isObjectIDFiltersOnly(params)) {
+		if (isSafeForUPASort(params)) {
+			if (params.getType().isPresent()) {
+				final TypeDefId t = params.getType().get();
+				// TODO CODE use optionals for typedefid versions
+				if (t.getMinorVersion() != null) {
+					sort.put(Fields.VER_TYPE_FULL, 1);
+				} else if (t.getMajorVersion() != null) {
+					sort.put(Fields.VER_TYPE_WITH_MAJOR_VERSION, 1);
+				} else {
+					sort.put(Fields.VER_TYPE_NAME, 1);
+				}
+			}
 			sort.put(Fields.VER_WS_ID, 1);
 			sort.put(Fields.VER_ID, 1);
 			sort.put(Fields.VER_VER, -1);
 		}
 		return sort;
 	}
 	
-	/* Check if there are no filters set other than the object ID filters. The object ID filters
-	 * may or may not be set.
-	 * The other filters are the two date filters, the metadata filter, the savers filter, and
-	 * the type filter.
+	/* Check if there are no filters set other than the object ID and type filters. The object ID
+	 * and type filters may or may not be set.
+	 * The other filters are the two date filters, the metadata filter, and the savers filter.
 	 * @return true if no filters other than the object ID filters are set.
 	 */
-	private boolean isObjectIDFiltersOnly(final ResolvedListObjectParameters params) {
+	private boolean isSafeForUPASort(final ResolvedListObjectParameters params) {
 		// be really careful about modifying this function. See notes in this file.
-		boolean oidFiltersOnly = !params.getAfter().isPresent()
+		return !params.getAfter().isPresent()
 				&& !params.getBefore().isPresent()
 				&& params.getMetadata().isEmpty()
-				&& params.getSavers().isEmpty()
-				// must have workspaces specified
-				&& !params.getType().isPresent();
-		return oidFiltersOnly;
+				&& params.getSavers().isEmpty();
 	}
 
 	private DBObject buildQuery(final ResolvedListObjectParameters params) {
-		final Set<Long> ids = params.getPermissionSet().getWorkspaces().stream()
-				.map(ws -> ws.getID()).collect(Collectors.toSet());
+		final List<Long> ids = params.getPermissionSet().getWorkspaces().stream()
+				.map(ws -> ws.getID()).distinct().sorted().collect(Collectors.toList());
```

Comment:
Since the query is passed to the ObjectInfoUtils mock, my choices are 1) write a custom matcher which I know from experience is verbose, annoying, and hard to read, or 2) sort here so the resulting list is deterministic. Since MongoDB will sort `$in` lists anyway, this moves a trivial amount of load from the db to this application for the cost of a trivial amount of memory, so win / win, if not very large wins.

The workspaces in the permission set *should* be unique, but looking at the code that's not guaranteed, hence the `distinct()` call.